lib/AST/ProtocolConformance.cpp

```diff
@@ -317,6 +317,109 @@ void NormalProtocolConformance::setTypeWitness(
   TypeWitnesses[assocType] = std::make_pair(substitution, typeDecl);
 }
 
+#ifndef NDEBUG
+static bool isSelfParameter(CanType type) {
+  if (auto param = dyn_cast<GenericTypeParamType>(type))
+    return param->getDepth() == 0 && param->getIndex() == 0;
+  return false;
+}
+#endif
+
+/// TypeWitnesses is keyed by the protocol's own declarations, but
+/// DependentMemberTypes will sometimes store a base protocol's declaration.
+/// Map to the derived declaration if possible.
+static AssociatedTypeDecl *getOwnAssociatedTypeDecl(ProtocolDecl *protocol,
+                                                    AssociatedTypeDecl *assoc) {
+  // Fast path.
+  if (assoc->getProtocol() == protocol) return assoc;
+
+  // Search the protocol.
+  for (auto member : protocol->getMembers()) {
+    if (auto memberAssoc = dyn_cast<AssociatedTypeDecl>(member)) {
```

Comment:
They won't be used as keys in the TypeWitnesses map, so I don't think that's necessary.  This is really just a temporary workaround around the bad internal representation of NormalProtocolConformance, which will need to be improved to handle generalized requirements anyway — it's literally not even storing conformance references for associated paths of length > 1.